components/templates/src/global_fns/load_data.rs

```diff
@@ -0,0 +1,346 @@
+            let file_time = get_file_time(&path).expect("get file time");
+            let file_datetime: DateTime<Utc> = DateTime::from(file_time);
+            format!("{}{}", file_datetime.timestamp_millis().to_string(), path.display())
+        }
+    };
+    return hex_digest(Algorithm::MD5, format!("{}{}", format, content_based_data).as_bytes());
+}
+
+
+fn get_data_from_args(content_path: &PathBuf, args: &HashMap<String, Value>) -> Result<ProvidedArgument> {
+    let path_arg = optional_arg!(
+        String,
+        args.get("path"),
+        GET_DATA_ARGUMENT_ERROR_MESSAGE
+    );
+
+    let url_arg = optional_arg!(
+        String,
+        args.get("url"),
+        GET_DATA_ARGUMENT_ERROR_MESSAGE
+    );
+
+    if !path_arg.is_some().bitxor(url_arg.is_some()) {
+        return Err("GET_DATA_ARGUMENT_ERROR_MESSAGE.into()".into());
+    }
+
+    if let Some(path) = path_arg {
+        let full_path = content_path.join(path);
+        if !full_path.exists() {
+            return Err(format!("{} doesn't exist", full_path.display()).into());
+        }
+        return Ok(ProvidedArgument::PATH(full_path));
+    }
+    else if let Some(url) = url_arg {
+        return Url::parse(&url).map(|parsed_url| ProvidedArgument::URL(parsed_url)).map_err(|e| format!("Failed to parse {} as url: {}", url, e).into());
+    }
+
+    return Err(GET_DATA_ARGUMENT_ERROR_MESSAGE.into());
+}
+
+fn read_data_file(base_path: &PathBuf, full_path: PathBuf) -> Result<String> {
+    if !is_path_in_directory(&base_path, &full_path).map_err(|e| format!("Failed to read data file {}: {}", full_path.display(), e))? {
+        return Err(format!("{} is not inside the base site directory {}", full_path.display(), base_path.display()).into());
+    }
+    return read_file(&full_path)
+        .map_err(|e| format!("`load_data`: error {} loading file {}", full_path.to_str().unwrap(), e).into());
+}
+
+fn get_output_format_from_args(args: &HashMap<String, Value>, provided_argument: &ProvidedArgument) -> Result<String> {
+    let format_arg = optional_arg!(
+        String,
+        args.get("format"),
+        "`load_data`: `format` needs to be an argument with a string value, being one of the supported `load_data` file types (csv, json, toml)"
+    );
+
+    if let Some(format) = format_arg {
+        return Ok(format);
+    }
+    return match provided_argument {
+        ProvidedArgument::PATH(path) => path.extension().map(|extension| extension.to_str().unwrap().to_string()).ok_or(format!("Could not determine format for {} from extension", path.display()).into()),
+        _ => Ok(String::from("plain"))
+    }
+}
+
+
+/// A global function to load data from a data file.
+/// Currently the supported formats are json, toml and csv
+pub fn make_load_data(content_path: PathBuf, base_path: PathBuf) -> GlobalFn {
+    let mut headers = header::HeaderMap::new();
+    headers.insert(header::USER_AGENT, format!("{} {}", crate_name!(), crate_version!()).parse().unwrap());
```

Comment:
is clap added to the dependencies just for that? I would just hardcode `zola` and not have the version imo, it's not that important.